File: 2226_candies.cpp
#include <iostream>

using namespace std;

bool is_valid(int guess, vector<int> &candies, long long k) {
  int count = 0;
  for (auto candy : candies) {
    count += candy / guess;
  }
  return count >= k;
}

int solve(vector<int> candies, long long k) {
  int ans = 0;
  int left = 0;
  int right = *max_element(candies.begin(), candies.end());
  while (left <= right) {
    int mid = left + (right - left) / 2;
    if (is_valid(mid, candies, k)) {
      ans = mid;
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return ans;
}

int main() {
  cout << solve({5, 8, 6}, 3) << endl;

  cout << solve({2, 5}, 11) << endl;

  return 0;
}

File: 2226_candies.py
def solve(candies, k):
    def assign_candies(piles, candy, k):
        if candy == 0:
            return True
        count = 0
        for pile in piles:
            count += pile // candy
            if count >= k:
                return True
        return False

    ans = None
    left, right = 0, sum(candies)//k
    while left <= right:
        mid = left + (right - left) // 2
        if assign_candies(candies, mid, k):
            if ans is None:
                ans = mid
            ans = max(ans, mid)
            left = mid + 1
        else:
            right = mid - 1

    return ans


print(solve([5, 8, 6], 3))
print(solve([2, 5], 11))
